### PR TITLE
Create a basic ECS

### PR DESCRIPTION
**Pull request by mathiasmagnusson**
First parts before closing #5 

### Features to implement before merging this PR into main

- [x] Keep track of entities
- [x] Store components
- [x] Query for (multiple) components
- [x] Iterate over all entities matching a query
- [x] Resources

### CODE REVIEW

**Comment by ollisco** [comment]
We could maybe split up workflows so that everything is not in the same file.

File: ecs/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+#[cfg(test)]
+mod tests {
+    use std::{any, collections::HashSet};
+
+    use crate::component::{ComponentRegistry, Storage, StorageType};
+
+    use super::*;
+
+    #[test]
+    fn component_registry() {
```
Comment (ollisco) [inline ecs/src/lib.rs:19]: Maybe add some documentation on the tests to clear up what purpose they serve

File: ecs/src/component/registry.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+impl ComponentInfo {
+    pub fn name(&self) -> &str {
+        self.name.as_ref()
+    }
+}
+
+#[derive(Debug)]
+pub struct ComponentEntry {
```
Comment (ollisco) [inline ecs/src/component/registry.rs:25]: Maybe som documentation

**Review by ollisco — COMMENTED** [review]
Overall great just a few comments on what could be improved slightly.

Some documentation of the more complex features and on the tests.

File: ecs/src/lib.rs (hunk 1 above)
Comment (mathiasmagnusson) [reply ecs/src/lib.rs:21]: Most of them kind of test what they say they do.

Perhaps the issue is mostly that the tests are too big and generic and should be broken up in smaller, more specific, test?

Comment (ollisco) [reply ecs/src/lib.rs:21]: Yes, maybe it is just me who don't understand the ECS code fully so understanding their tests are more difficult

**Comment by mathiasmagnusson** [comment]
Retrieving data is pretty unergonomic and requires unsafe code (since what you get is raw pointers).

I think the general solution would require proc macros and I don't think that's worth the hassle TBF.

**Review by ollisco — DISMISSED** [review]
Looks good. 

I agree that we don't have to prioritise removing unsafe even though it would be nice :)

**Review by ghagl — DISMISSED** [review]
Ser bra ut. Härnäst blir det att fundera över föräldra-barn-relationer (d.v.s. om ett svärd finns hos en karaktär, tas karaktären bort så ska svärdet bort)

**Comment by ollisco** [comment]
@mathiasmagnusson Kan du fixa dessa merge conflicts?

**Comment by mathiasmagnusson** [comment]
Yes jag fixat det idag

**Comment by mathiasmagnusson** [comment]
Seems like we need those reviews again...
The only change is that i merged changes from the main branch

**Review by ollisco — APPROVED** [review]
looks good

**Review by ghagl — APPROVED** [review]

**Pull request merged**